### ISSUE TITLE
[BUG] Discrepancy with HORTON's cart-sph transformation matrix

### ISSUE THREAD

**Issue by kimt33** [theochem/gbasis#89]
**Describe the bug**
The generated transformation matrix for cartesian to spherical contractions/primitives is different from the one from HORTON (https://github.com/theochem/horton/blob/master/horton/gbasis/cartpure.cpp) for angular momentum 4 and 5 (and probably all of the higher angular momentums).

In the case of angular momentum 4, `gbasis`'s transformation matrix contains an extra element at position (3, 3), which corresponds to C_{42} (or m=2) and X(xxyy) (or `(2, 2, 0)`).

In the case of angular momentum 5, `gbasis`'s transformation matrix contains extra elements at positions (3, 7), which correspond to C_{52} (m=2) and X(xxyyz) (`(2, 2, 1)`) and (6,6) which corresponds to S_{53} (m=-3) and X(xxyyy) (`(2, 3, 0)`). `horton`'s transformation matrix contains an extra element at position (5, 3) which corresponds to C_{53} (m=3) and X(xxxyy) (`(3, 2, 0)`).

**To Reproduce**
Follow code was used to create the HORTON's transformation matrix:
```python
import itertools as it

from gbasis.spherical import generate_transformation
import numpy as np


# taken from HORTON's gbasis/cartpure.cpp
horton_transform = [
    [0,  0, 0.375],
    [0,  3, 0.21957751641341996535],
    [0,  5, -0.87831006565367986142],
    [0, 10, 0.375],
    [0, 12, -0.87831006565367986142],
    [0, 14, 1.0],
    [1,  2, -0.89642145700079522998],
    [1,  7, -0.40089186286863657703],
    [1,  9, 1.19522860933439364],
    [2,  4, -0.40089186286863657703],
    [2, 11, -0.89642145700079522998],
    [2, 13, 1.19522860933439364],
    [3,  0, -0.5590169943749474241],
    [3,  5, 0.9819805060619657157],
    [3, 10, 0.5590169943749474241],
    [3, 12, -0.9819805060619657157],
    [4,  1, -0.42257712736425828875],
    [4,  6, -0.42257712736425828875],
    [4,  8, 1.1338934190276816816],
    [5,  2, 0.790569415042094833],
    [5,  7, -1.0606601717798212866],
    [6,  4, 1.0606601717798212866],
    [6, 11, -0.790569415042094833],
    [7,  0, 0.73950997288745200532],
    [7,  3, -1.2990381056766579701],
    [7, 10, 0.73950997288745200532],
    [8,  1, 1.1180339887498948482],
    [8,  6, -1.1180339887498948482],
]
answer = np.zeros((9, 15))
for i in horton_transform:
    answer[i[0], i[1]] = i[2]
test = (
    np.abs(
        generate_transformation(
            4,
            np.array(
                [
                    (i.count(0), i.count(1), i.count(2))
                    for i in it.combinations_with_replacement(range(3), 4)
                ]
            ),
            (0, 1, -1, 2, -2, 3, -3, 4, -4),
            "left",
        ),
    )
    - np.abs(answer)
)
print(
    np.where((test > 1e-9))
)
print(
    np.where((test < -1e-9))
)

# taken from HORTON's gbasis/cartpure.cpp
horton_transform = [
    [0,  2, 0.625],
    [0,  7, 0.36596252735569994226],
    [0,  9, -1.0910894511799619063],
    [0, 16, 0.625],
    [0, 18, -1.0910894511799619063],
    [0, 20, 1.0],
    [1,  0, 0.48412291827592711065],
    [1,  3, 0.21128856368212914438],
    [1,  5, -1.2677313820927748663],
    [1, 10, 0.16137430609197570355],
    [1, 12, -0.56694670951384084082],
    [1, 14, 1.2909944487358056284],
    [2,  1, 0.16137430609197570355],
    [2,  6, 0.21128856368212914438],
    [2,  8, -0.56694670951384084082],
    [2, 15, 0.48412291827592711065],
    [2, 17, -1.2677313820927748663],
    [2, 19, 1.2909944487358056284],
    [3,  2, -0.85391256382996653193],
    [3,  9, 1.1180339887498948482],
    [3, 16, 0.85391256382996653193],
    [3, 18, -1.1180339887498948482],
    [4,  4, -0.6454972243679028142],
    [4, 11, -0.6454972243679028142],
    [4, 13, 1.2909944487358056284],
    [5,  0, -0.52291251658379721749],
    [5,  3, 0.22821773229381921394],
    [5,  5, 0.91287092917527685576],
    [5, 10, 0.52291251658379721749],
    [5, 12, -1.2247448713915890491],
    [6,  1, -0.52291251658379721749],
    [6,  6, -0.22821773229381921394],
    [6,  8, 1.2247448713915890491],
    [6, 15, 0.52291251658379721749],
    [6, 17, -0.91287092917527685576],
    [7,  2, 0.73950997288745200532],
    [7,  7, -1.2990381056766579701],
    [7, 16, 0.73950997288745200532],
    [8,  4, 1.1180339887498948482],
    [8, 11, -1.1180339887498948482],
    [9,  0, 0.7015607600201140098],
    [9,  3, -1.5309310892394863114],
    [9, 10, 1.169267933366856683],
    [10,  1, 1.169267933366856683],
    [10,  6, -1.5309310892394863114],
    [10, 15, 0.7015607600201140098],
]
answer = np.zeros((11, 21))
for i in horton_transform:
    answer[i[0], i[1]] = i[2]

test = (
    np.abs(
        generate_transformation(
            5,
            np.array(
                [
                    (i.count(0), i.count(1), i.count(2))
                    for i in it.combinations_with_replacement(range(3), 5)
                ]
            ),
            (0, 1, -1, 2, -2, 3, -3, 4, -4, 5, -5),
            "left",
        ),
    )
    - np.abs(answer)
)
print(
    np.where((test > 1e-9))
)
print(
    np.where((test < -1e-9))
)
```

**Expected behaviour**
They are expected to be the same.

**System Information:**
 - OS: Fedora 26
 - Python version: Python 3.6.8
 - NumPy version: 1.16.4
 - SciPy version: 1.3.0

**Additional context**
I'm looking into this, but could @wilhadams check the `gbasis.spherical` part of the code for bugs and @tovrstra verify that the HORTON's transformation matrix is correct? It'd also be great if we are able to compare against another reference.

**Comment by wilhadams** [comment]
There appears to be an error in `real_solid_harmonic`, C_{4,2} is not calculated correctly, comparing to Helgaker as a reference. Helgaker is in agreement with the [Horton docs](https://theochem.github.io/horton/2.1.1/tech_ref_gaussian_basis.html?highlight=cartpure).

**Comment by PaulWAyers** [comment]
C_{4,2} is wrong but S_{4,2} is right? That seems very strange....I wouldn't expect the formula to work for one and not the other.

**Comment by wilhadams** [comment]
I've tracked down the real issue. When a term appears multiple times (in this case x^2y^2), the terms aren't cancelling out correctly. All the math is working correctly, just a failure to assemble the final coefficients correctly.

**Comment by PaulWAyers** [comment]
This is something I know we talked about before, so maybe David remembers. I just double-checked the HORTON result and it seems plausible to me now; before I was skeptical because there is a strange "asymmetry" but this seems OK in retrospect.

**Comment by PaulWAyers** [comment]
That makes sense Wil. There are several ways to write the equations, depending on whether one factors out the r^2 term or not and how one groups terms.....but it shouldn't affect the *numbers* computed, right?

**Comment by wilhadams** [comment]
This is what was computed for `real_spherical_harmonic(4, 2)`:

`i   j  k   || x  y  z ||           coeff`
`0 0 0.0 || 2 0 2 || 3.3541019662496847`
`0 0 1.0 || 0 2 2 || -3.3541019662496847`
`1 0 0.0 || 4 0 0 || -0.5590169943749475`
`1 0 1.0 || 2 2 0 || 0.5590169943749475`
`1 1 0.0 || 2 2 0 || -0.5590169943749475`
`1 1 1.0 || 0 4 0 || 0.5590169943749475`

So yes, basically just the only one the program was grabbing was the negative version, which came up second in the list.

This is fixed in PR #95
